Tasks/Gradle/testlib/testlib-main.ts

```diff
@@ -0,0 +1,416 @@
+export function rmRF(path:string, continueOnError?:boolean):void {
+    debug('rm -rf ' + path);
+    shell.rm('-rf', path);
+
+    var errMsg:string = shell.error();
+
+    // if you try to delete a file that doesn't exist, desired result is achieved
+    // other errors are valid
+    if (errMsg && !(errMsg.indexOf('ENOENT') === 0)) {
+        throw new Error('rm -rf failed with error ' + errMsg);
+    }
+}
+
+// A working implementation of tl.checkPath().
+// Warning: fully functional. If the file does not exist, the task will fail.
+export function checkPath(p:string, name:string):void {
+    debug('check path : ' + p);
+    if (!p) {
+
+        setResult(TaskResult.Failed, 'not found ' + name + ': ' + p);  // exit
+    }
+}
+
+// A working implementation of tl.mkdirP().
+// Warning: fully functional, unlike the test implementation of tl.mkdirP()
+export function mkdirP(p):void {
+    if (!p) {
+        throw new Error('Expected a parameter.');
+    }
+
+    // certain chars like \0 will cause shelljs and fs
+    // to blow up without exception or error
+    if (p.indexOf('\0') >= 0) {
+        throw new Error(loc('LIB_PathHasNullByte'));
+    }
+
+    if (!shell.test('-d', p)) {
+        debug('creating path: ' + p);
+        shell.mkdir('-p', p);
+        checkShell('mkdirP');
+    }
+    else {
+        debug('path exists: ' + p);
+    }
+}
+
+// A working implementation of tl.cp().
+// Warning: fully functional, unlike the test implementation of tl.cp()
+export function cp(source:string, dest:string, options?:string, continueOnError?:boolean):void {
+    if (options) {
+        shell.cp(options, source, dest);
+    }
+    else {
+        shell.cp(source, dest);
+    }
+
+    checkShell('cp', continueOnError);
+}
+
+// A working implementation of tl.exist().
+// Warning: fully functional, unlike the test implementation of tl.exist()
+export function exist(path:string):boolean {
+    var exist = false;
+    try {
+        exist = path && fs.statSync(path) != null;
+    } catch (err) {
+        if (err && err.code === 'ENOENT') {
+            exist = false;
+        } else {
+            throw err;
+        }
+    }
+    return exist;
+}
+
+// Working implementation of setResourcePath(path).
+// Warning: multiple calls to this file will result in odd behaviour.
+export function setResourcePath(path:string):void {
+    if (!resourceFilePath) {
+        resourceFilePath = path;
+        debug('set resource file to: ' + resourceFilePath);
+
+        var locStrs = loadLocStrings(resourceFilePath);
+        for (var key in locStrs) {
+            debug('cache loc string: ' + key);
+            setLoc(key, locStrs[key]);
+        }
+
+    }
+    else {
+        warning('resource file is already set to: ' + resourceFilePath);
+    }
+}
+
+// Replaces tl.loc(). Use setResourcePath(path) or setLoc(key, value) in setup.
+export function loc(key:string):string {
+    return 'Localisation of key ' + key;
```

Comment:
To change: this is not a functional implementation of _loc()_. There is a problem with the application of _tl.loc()_ here, which normally is of the format _tl.loc(locKey, params...)_ where _params_ are used as String.Format()-style replacements. However, due to incompabilities with the compilation, we cannot use params here.

Will change to a simple lookup without replacement.